engine/src/test/java/io/camunda/zeebe/streamprocessor/StreamProcessorReplayTest.java

```diff
@@ -0,0 +1,266 @@
+import org.assertj.core.api.Assertions;
+import org.awaitility.Awaitility;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.InOrder;
+import org.mockito.Mockito;
+import org.mockito.verification.VerificationWithTimeout;
+
+@ExtendWith(StreamPlatformExtension.class)
+public final class StreamProcessorReplayTest {
+
+  private static final long TIMEOUT_MILLIS = 2_000L;
+  private static final VerificationWithTimeout TIMEOUT = timeout(TIMEOUT_MILLIS);
+
+  private static final ProcessInstanceRecord RECORD = Records.processInstance(1);
+
+  @SuppressWarnings("unused") // injected by the extension
+  private StreamPlatform streamPlatform;
+
+  @Test
+  public void shouldReplayEvents() {
+    // given
+    streamPlatform.writeBatch(
+        command().processInstance(ACTIVATE_ELEMENT, RECORD),
+        event().processInstance(ELEMENT_ACTIVATING, RECORD).causedBy(0));
+
+    // when
+    streamPlatform.startStreamProcessor();
+
+    // then
+    final RecordProcessor recordProcessor = streamPlatform.getDefaultMockedRecordProcessor();
+    final InOrder inOrder = inOrder(recordProcessor);
+    inOrder.verify(recordProcessor, TIMEOUT).replay(any());
+    inOrder.verifyNoMoreInteractions();
+  }
+
+  @Test
+  public void shouldProcessAfterReplay() {
+    // given
+    streamPlatform.writeBatch(
+        command().processInstance(ACTIVATE_ELEMENT, RECORD),
+        event().processInstance(ELEMENT_ACTIVATING, RECORD).causedBy(0));
+
+    // when
+    streamPlatform.startStreamProcessor();
+
+    streamPlatform.writeBatch(
+        command().processInstance(ACTIVATE_ELEMENT, RECORD),
+        event().processInstance(ELEMENT_ACTIVATING, RECORD).causedBy(0));
+
+    // then
+    final RecordProcessor recordProcessor = streamPlatform.getDefaultMockedRecordProcessor();
+    final InOrder inOrder = inOrder(recordProcessor);
+    inOrder.verify(recordProcessor, TIMEOUT).replay(any());
+    inOrder.verify(recordProcessor, TIMEOUT).process(any(), any());
+    inOrder.verifyNoMoreInteractions();
+  }
+
+  @Test
+  public void shouldSkipCommands() {
+    // given
+    streamPlatform.writeBatch(
+        command().processInstance(ACTIVATE_ELEMENT, RECORD),
+        event().processInstance(ELEMENT_ACTIVATING, RECORD).causedBy(0));
+
+    // when
+    streamPlatform.startStreamProcessor();
+
+    // then
+    final RecordProcessor recordProcessor = streamPlatform.getDefaultMockedRecordProcessor();
+    final InOrder inOrder = inOrder(recordProcessor);
+    inOrder.verify(recordProcessor, TIMEOUT).replay(any());
+    inOrder.verifyNoMoreInteractions();
+  }
+
+  @Test
+  public void shouldSkipRejections() {
+    // given
+    streamPlatform.writeBatch(
+        command().processInstance(ACTIVATE_ELEMENT, RECORD),
+        rejection().processInstance(ACTIVATE_ELEMENT, RECORD).causedBy(0));
+
+    // when
+    streamPlatform.startStreamProcessor();
```

Comment:
This is because we have on the log a command which caused a rejection. This means this command is already processed. During replay we will not read the commands nor the rejection, we are only accepting events and processing should also not accept any of the existing records. This is what we want to verify here.